added previous-url to global-state

ike-kottlowski opened the pull request:
## Type of change

<!-- (mark with an `X`) -->

```
- [x] Bug fix
- [ ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

<!--Describe what the purpose of this PR is. For example: what bug you're fixing or what new feature you're adding-->
PR addresses issue when user tries to deep link into web-vault using SSO login flow. 

## Code changes

<!--Explain the changes you've made to each file or major component. This should help the reviewer understand your changes-->
<!--Also refer to any related changes or PRs in other repositories-->

- **libs/common/src/platform/models/domain/global-state.ts :** added `previousUrl: string` to track deep link URL. Adding it to `global-state` means allows `state.service.ts` to store the `global-state` in local storage in the browser.
- **state.service.ts :** 
  - **Abstraction :** added `(get/set)PreviousUrl` methods
  - **Lib common Implementation :** implemented get/set
  - **Web Implementation :** called get/set from _Lib Common_ implementation
- **apps/web/src/app/core/router.service.ts :** Added state service and modified get/set methods to fetch from state.
- All other files were minor updates to account for `async` when calling the `getPreviousUrl` method.

## Screenshots

<!--Required for any UI changes. Delete if not applicable-->

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

Hinton reviewed (commented):
I think this might have unintented side-effects. If a user opens bitwarden by navigating to any page locked behind a login wall, and then closes the window. And at a later stage attempts to login regularly they would now navigate to that page since we persist the previous url.

I think we should only store the previous url in a persistent state if the user actively invokes a login method that sends them away from Bitwarden such as SSO. And only load it if they return from such a page. We should also set a null value if previous url is null, so that previous states aren't persisted.

We should also revert the previous fix for emergency access since that would now be unnecessary.

Hinton commented on apps/web/src/app/auth/login/login-with-device.component.ts:
```diff
@@ -60,5 +61,22 @@ export class LoginWithDeviceComponent
+  async goAfterLogIn() {
+    this.loginService.clearValues();
+    let previousUrl = await this.stateService.getPreviousUrl();
+    if (previousUrl == "" || previousUrl == undefined || previousUrl == null) {
```
You can use the helper function for this.

```suggestion
    if (!Utils.isNullOrEmpty(previousUrl)) {
```

ike-kottlowski replied:
That's the one I was thinking of. Thanks, it felt wrong to write it that way.

ike-kottlowski replied:
Fixed in most this [commit](https://github.com/bitwarden/clients/pull/5733/commits/a525c8066ddda02f47539f7ad1bc3be478bddc63.)

Hinton commented on apps/web/src/app/auth/login/login-with-device.component.ts:
```diff
@@ -60,5 +61,22 @@ export class LoginWithDeviceComponent
+      this.stateService.setPreviousUrl(null);
+      this.router.navigateByUrl(previousUrl);
+    } else {
+      this.router.navigate([this.successRoute], {
+        queryParams: {},
+      });
```
Is there a reason to specify `queryParams` here?

ike-kottlowski replied:
That there is some prime handmade _copy pasta_. I have removed it in [this commit](https://github.com/bitwarden/clients/pull/5733/commits/f2f1029a138369fe4edd91d004e820d939f5b139).

Thanks for catching it.  👍

Hinton commented on apps/web/src/app/auth/login/login.component.ts:
```diff
@@ -90,6 +90,8 @@ export class LoginComponent extends BaseLoginComponent implements OnInit, OnDest
   }
 
   async ngOnInit() {
+    this.stateService.setPreviousUrl(null);
```
Maybe add a comment to specify why we reset `stateService` here.

ike-kottlowski replied:
Added in this [commit](https://github.com/bitwarden/clients/pull/5733/commits/a525c8066ddda02f47539f7ad1bc3be478bddc63)

Hinton commented on apps/web/src/app/auth/sso.component.ts:
```diff
@@ -61,20 +64,9 @@ export class SsoComponent extends BaseSsoComponent {
-      };
-    }
+    this.onSuccessfulLoginNavigate = async () => {
+      this.router.navigateByUrl(await this.stateService.getPreviousUrl());
```
What happens if the url is empty?

ike-kottlowski replied:
When this URL is empty it will navigate to the default vault screen.

Hinton reviewed (commented):
I think this approach is sound. I would try and move the state calls to the router service, you could have three methods, `persistPreviousUrl()`, `clearPersistedUrl()`, `getPersistedUrl()` which better describes what's happening. And removes direct calls to `stateService` from the components which is a long term architectural vision.

ike-kottlowski commented:
> I think this approach is sound. I would try and move the state calls to the router service, you could have three methods, `persistPreviousUrl()`, `clearPersistedUrl()`, `getPersistedUrl()` which better describes what's happening. And removes direct calls to `stateService` from the components which is a long term architectural vision.

I updated the components to act in this way by placing the calls to `GlobalState` into the `RouterService`.

Hinton commented on apps/web/src/app/core/state/state.service.ts:
```diff
@@ -114,4 +114,14 @@ export class StateService extends BaseStateService<GlobalState, Account> {
     return await super.setLastSync(value, options);
   }
+
+  async getPreviousUrl(options?: StorageOptions): Promise<string> {
+    options = this.reconcileOptions(options, await this.defaultOnDiskLocalOptions());
+    return await super.getPreviousUrl(options);
+  }
+
+  async setPreviousUrl(url: string, options?: StorageOptions): Promise<void> {
+    options = this.reconcileOptions(options, await this.defaultOnDiskLocalOptions());
+    return await super.setPreviousUrl(url, options);
+  }
```
This seems to be identical to the default implementation?

ike-kottlowski replied:
yep this isn't needed.

Hinton commented on apps/web/src/app/auth/login/login.component.ts:
```diff
@@ -90,6 +90,13 @@ export class LoginComponent extends BaseLoginComponent implements OnInit, OnDest
 
   async ngOnInit() {
+    /**
+     * Set GlobalState.PreviousUrl to NULL onInit() to handle situation where
+     * a previous SSO login was interrupted. Ensures the user is navigated
+     * to the correct location.
```
This comment includes some internal implementation details of the `RouterService`. I think it's enough to just mention why we want to clear it, not how it's cleared. Something like:

```
Clear `persistedUrl` to handle flows where SSO login were interrupted. Ensures the user is navigated to the correct location on login.
```

Hinton commented on apps/web/src/app/auth/login/login-with-device.component.ts:
```diff
@@ -60,5 +62,20 @@ export class LoginWithDeviceComponent
+
+  async goAfterLogIn() {
+    this.loginService.clearValues();
+    let previousUrl = await this.routerService.getPersistedUrl();
+    if (Utils.isNullOrEmpty(previousUrl)) {
+      previousUrl = this.routerService.getPreviousUrl();
+    }
+    if (previousUrl) {
+      this.routerService.clearPersistedUrl();
```
You can lift this into the router service.

```ts
async getAndClearPersistedPreviousUrl(): Promise<void> {
  const persisted = await this.stateService.getPreviousUrl();

  if (!Utils.isNullOrEmpty(persisted)) {
    await this.clearPersistedUrl();
    return persisted;
  }

  return this.getPreviousUrl();
}
```

Which lets you simplify the code in this component and two-factor-component.

```ts
const previousUrl = await this.routerService.clearAndGetPersistedPreviousUrl();
this.router.navigateByUrl(Utils.isNullOrEmpty(previousUrl) ? [this.successRoute] : previousUrl);
```

Hinton reviewed (commented):
A comment that needs updating, and an optional refactor.

ike-kottlowski commented on libs/angular/src/auth/guards/redirect.guard.ts:
```diff
@@ -38,7 +41,13 @@ export function redirectGuard(overrides: Partial<RedirectRoutes> = {}): CanActiv
     }
 
     if (authStatus === AuthenticationStatus.Unlocked) {
```
I'm not sure about this implementation. I am able to deep link in by using the `StateService`, but I'm not sure I like using the `web/RouterService` to access the `PersistedUrl` and now we are also accessing it in here. Has maintainability-smells. But the `web/RouterService` cannot be referenced here so this was what I came up with short of re-writing the whole deep link fix as a guard.

Hinton replied:
Not sure I follow. This is a guard to redirect the user correct when landing on the home page. Why would we need to navigate from here?

I suspect this would have unintended side effects if you are on a page and then attempt to navigate directly to `/`.

ike-kottlowski replied:
The core issue trying to be solved is general deep link availability no matter which auth/decryption method being used. 

This specific guard will route a user to `/vault` endpoint after a successful auth/decryption flow. I am trying to address the user experience issue of navigating to a specific item when the user is in a logged-out/locked state and ultimately getting to the item the user wanted to navigate to in the first place. In this guard, the only option after a successful login/decryption is `/vault`, which means a user will never be able to deep link using TDE.

I just tested deep linking into the vault using SSO + TDE. I am automagically navigated to the cypher after login. I verified the `global.prevoulUrl` is `null` in local storage. I then try to navigate to `/` and it routes me to `/vault`. So, navigating to `/` doesn't seem to cause any issues. 

If I'm not understanding your question, please let me know.

Hinton replied:
Consider the following scenario, we have two open tabs, one logged in, one logged out.

Logged out, navigate to any path, begin SSO process, abort after going to the IDP website.
Logged in: Navigate to `/`, get redirected to the path you tried to access in the other tab.

ike-kottlowski replied:
I was able to fix this issue by changing the storage of the URL to session storage rather than local storage in the browser, so the data is scoped on a per tab basis.

ike-kottlowski commented:
Ready to go with TDE updates. Deep linking and emergency access works for these flows: MP, Login with Device, SSO, MP + MFA, Login with Device + MFA, SSO + MFA, and SSO with TDE enabled.

Hinton commented on apps/browser/src/auth/popup/two-factor.component.ts:
Any reason this file was changed?

ike-kottlowski replied:
To support the base `two-factor.component.ts` I had to add the `stateService` back. the TDE merge required it to be injected.

Hinton replied:
Yes I understand why we need the dependency, what I was referring to is why wasn't the file changes reset. Unnecessary diffs should be avoided as it obfuscates the history of a file and makes it harder to later run git blame on them.

Hinton commented on apps/desktop/src/auth/two-factor.component.ts:
Any reason this file was changed?

ike-kottlowski replied:
To support the base `two-factor.component.ts` I had to add the `stateService` back. the TDE merge required it to be injected.

Hinton commented on apps/web/src/app/core/router.service.ts:
```diff
@@ -51,11 +54,50 @@ export class RouterService {
   }
 
-  getPreviousUrl() {
+  getPreviousUrl(): string {
```
These can return undefined, so I think a more accurate type is to include undefined.

```suggestion
  getPreviousUrl(): string | undefined {
```

Hinton commented on apps/web/src/app/core/router.service.ts:
```diff
@@ -51,11 +54,50 @@ export class RouterService {
+   * Fetch GlobalState.PreviousUrl
+   * @returns GlobalState.PreviousUrl
+   */
+  async getPersistedUrl(): Promise<string> {
+    return await this.stateService.getPreviousUrl();
+  }
```
```suggestion
  async getPersistedUrl(): Promise<string | undefined> {
    return await this.stateService.getPreviousUrl();
  }
```

Hinton commented on apps/web/src/app/core/router.service.ts:
```diff
@@ -51,11 +54,50 @@ export class RouterService {
+   * Fetch and clear persisted URL if present in state, otherwise return
+   * router `previousUrl`
+   */
+  async getAndClearPersistedPreviousUrl(): Promise<string> {
```
```suggestion
  async getAndClearPersistedPreviousUrl(): Promise<string | undefined> {
```

Hinton commented on libs/angular/src/auth/components/login-with-device.component.ts:
I don't think this file needs to change?

Hinton commented on libs/angular/src/auth/components/two-factor.component.ts:
Any reason this file was changed?

Hinton reviewed (commented):
There seems to have been accidental changes to a number of files, to keep the history cleaner we should revert those.

Still a bit unsure if the redirect guard won't have unintentional side effects. The lock component already tries to navigate you to the previousUrl in the router.

ike-kottlowski commented:
Change of direction from initial idea. After various conversations with @JaredSnider-Bitwarden the guard seems like the most elegant solution.

Fixed the multi-tab navigation issue by changing the storage location to session rather than local storage. 

I tested deep linking to the vault, and organization modules as well as accepting emergency access.

gbubemismith commented:
@Hinton, I think this #5989 would not be needed anymore when this goes through

Hinton commented:
@gbubemismith this is for the web vault, unless the scope has changed.

Hinton reviewed (changes requested):
A few questions.

- Aren't we losing regular deep linking capability since login never checks the previous url?
- We seem to always "deep persist" links. But still retain support for previousUrl, previousUrl is only used for login checks.
- It's a bit confusing that we have to ways of redirecting users now.

Hinton commented on apps/web/src/app/auth/login/login.component.ts:
```diff
@@ -90,6 +90,12 @@ export class LoginComponent extends BaseLoginComponent implements OnInit, OnDest
+     * clear `persistedUrl` to handle flows when SSO login were interrupted.
+     * Ensures the user is navigated to the correct location.
+     */
+    this.routerService.clearPersistedPreLoginDeepLinkUrl();
```
Won't this clear the deep link before we even begin?

ike-kottlowski commented:
In this most recent commit, I opted to leave the `routerService.previousUrl` logic intact since the lock component need context on the `previousUrl` so the `pre-login-redirect.guard.ts` can store where the user was prior to the session being locked.

There is also various refactoring to increase readability.

ike-kottlowski commented:
I added test for the guards to increase our code coverage as well as added some comments.

jlf0dev commented on apps/web/src/app/core/guards/deep-link-cache.guard.spec.ts:
Should be moved under auth folder

jlf0dev commented on apps/web/src/app/core/guards/deep-link-cache.guard.spec.ts:
```diff
@@ -0,0 +1,134 @@
+  });
+
+  // Story: User logs in and does not try to deep link
+  // TODO Does the user get routed to vault twice when logging in without a deeplink?
```
Do we need to address this before merge?

ike-kottlowski replied:
I think I'm over thinking it. This test case is the same as the `currentUrl === undeinfed/null/""` test, I'll remove the comment.

jlf0dev commented on apps/web/src/app/core/guards/deep-link-cache.guard.ts:
\+ move to Auth folder

jlf0dev commented on apps/web/src/app/core/guards/deep-link-redirect.guard.spec.ts:
\+ move to Auth folder

jlf0dev commented on apps/web/src/app/core/guards/deep-link-redirect.guard.ts:
\+ move to Auth folder

jlf0dev reviewed (changes requested):
Fantastic tests, thank you for creating those! Just need to move some files under Auth ownership

JaredSnider-Bitwarden commented on apps/web/src/app/core/guards/deep-link-cache.guard.spec.ts:
TIL how to test guards neatly in Jest! This is awesome!

JaredSnider-Bitwarden commented on apps/web/src/app/core/guards/deep-link-cache.guard.spec.ts:
```diff
@@ -0,0 +1,134 @@
+    routerService.getPreviousUrl.mockReturnValue("/previous-url");
+
+    // Act
+    await routerHarness.navigateByUrl("/guarded-route");
```
I'd not used the `routerTestingHarness` before so I was reading about it [here](https://angular.io/api/router/testing/RouterTestingHarness#navigateByUrl), and I noticed that the `navigateByUrl(...)` returns a component instance or null based on if the route gets activated or not. 

This opens up additional assertion opportunities: 
```TypeScript
const activatedComponent = await routerHarness.navigateByUrl("/guarded-route");

// Assert
expect(activatedComponent).not.toBeNull(); // This tests if the route was activated.
expect(activatedComponent).toBeNull(); // This tests if the route was NOT activated.
```

ike-kottlowski replied:
Excellent, I'll add them in to get more verbose coverage.

ike-kottlowski commented:
I updated the `deep-link-cache-guard` to handle deep linking from the lock page. Added a test to verify the added functionality.

Added `expect(activatedComponent)` to the `deep-link-redirect-guard.spec.ts` to validate that the redirect component has been navigated to.

JaredSnider-Bitwarden reviewed (dismissed):
Excellent work on this! Thank you for the recordings demonstrating the various scenarios as well!

jlf0dev reviewed (approved):
Great implementation! I really like how this turned out 🚀

gbubemismith reviewed (dismissed):
Nice job @ike-kottlowski

eliykat reviewed (approved):
Approving AC Team changes, which are just adding the `deepLinkCacheGuard` to our routes. I haven't reviewed the rest of the PR but note that others have approved the guards and the overall approach.

Might be good to see if @Hinton wants to take another look before merge though.

Hinton commented on apps/web/src/app/core/router.service.ts:
```diff
@@ -51,11 +54,33 @@ export class RouterService {
+  /**
+   * Fetch and clear persisted LoginRedirectUrl if present in state
+   */
+  async getAndClearLoginRedirectUrl(): Promise<string> | undefined {
```
**question**: This only fetches the state persisted value, is there a point in having the `previousUrl` on routerService?

ike-kottlowski replied:
Yes, we require both. The `previousUrl` handles when a user's vault times out, and they are changed to the `lock` status. 

Since the `deep-link-cache-guard` only handles navigation prior to being `unlocked`, we need the `previousUrl` as it tracks navigation while the user is `unlocked`. When the vault times out we persist the 'previousUrl' to state. So, when a user is navigated to the `success-route` in the lock component the `deep-link-redirect-guard` will pull the URL from state and navigate the user to where they were prior to the vault timeout.

Hinton commented on apps/web/src/app/vault/individual-vault/vault-routing.module.ts:
```diff
@@ -1,12 +1,15 @@
     path: "",
     component: VaultComponent,
     data: { titleId: "vaults" },
+    canActivate: [deepLinkRedirectGuard()],
```
**question**: Is this needed? We already have a canActivate rule in `oss-routing` which this is nested under

Hinton reviewed (commented):
Two questions. It's a bit confusing having `previousUrl` and `persistLoginRedirectUrl` and I believe there is a unecessary canActivate.

ike-kottlowski commented:
I combined the two guards into one as @Hinton pointed out that we don't need the `canActivate` in the `vault-routing.module.ts`. Combining makes the purpose of the guard more straight forward easing maintenance requirements.

JaredSnider-Bitwarden reviewed (dismissed):
🚀 🎸 👍

eliykat commented on apps/web/src/app/oss-routing.module.ts:
```diff
@@ -20,7 +20,7 @@ import { CreateOrganizationComponent } from "./admin-console/settings/create-org
 import { AcceptEmergencyComponent } from "./auth/accept-emergency.component";
 import { AcceptOrganizationComponent } from "./auth/accept-organization.component";
-import { deepLinkCacheGuard } from "./auth/guards/deep-link-cache.guard";
+import { deepLinkGuard as deepLinkGuard } from "./auth/guards/deep-link.guard";
```
Unnecessary aliasing

eliykat reviewed (dismissed):
Very minor feedback, otherwise lgtm. Anyone from AC Team can feel free to blindly approve in my absence assuming the only further change is to remove the alias.

JaredSnider-Bitwarden reviewed (dismissed):
🚀

gbubemismith reviewed (approved):
LGTM!

ike-kottlowski commented:
merged master into branch. no new changes.

JaredSnider-Bitwarden reviewed (dismissed):
LGTM

JaredSnider-Bitwarden reviewed (approved):
LGTM!
